lib/stagger/tcp_client.js

```diff
@@ -0,0 +1,137 @@
+var events = require("events");
+var net = require("net");
+var util = require("util");
+
+var Parser = require('../protocol/parser');
+var encode = require('../protocol/encoder');
+
+// Manages a TCP connection to stagger, including reconnects
+// (De)serialises incoming/outgoing messages
+function TcpClient(tags, connectionOpts, timeout) {
+  events.EventEmitter.call(this);
+
+  this.timeout = timeout || 60000;
+  this.connected = false;
+
+  this.parser = new Parser(
+    function(method, body) {
+      switch (method) {
+        case 'pair:ping':
+          this.pong();
+          break;
+        case 'pair:pong':
+          break; // NOOP
+        default:
+          this.emit("message", method, body);
+          break;
+      }
+    }.bind(this),
+    function(err) {
+      console.log("Parse error: " + err);
+      this.socket.end(); // will trigger a close event, which will lead to a reconnect
+      this.handleShutdown();
+      this.parser.reset();
+    }.bind(this)
+  );
+
+  this.connect(tags, connectionOpts);
+}
+
+util.inherits(TcpClient, events.EventEmitter);
+var prototype = TcpClient.prototype;
+
+prototype.send = function(method, body) {
+  this.socket.write(encode(method, body));
```

Comment:
What do you think the behaviour should be if it's not connected? Just discards the messages, or keep buffering them and send them all once it reconnects (obviously there will have to be some limit on the number that are buffered).